text/0000-nested-method-calls.md

```diff
@@ -0,0 +1,824 @@
+the regions more explicitly:
+
+```rust
+/* 0 */ let mut i = 0;
+/* 1 */ let p: RefMut<{2-5}, {3,5}, i32> = &mut i;
+//                    ^^^^^  ^^^^^
+//                     'r     'w
+/* 2 */ let j = i;  // just in 'r
+/* 3 */ *p += 1;    // must be in 'w
+/* 4 */ let k = i;  // just in 'r
+/* 5 */ *p += 1;    // must be in 'w
+```
+
+As you can see here, we would infer the write region to be just the
+two points 3 and 5. This is precisely those portions of the CFG where
+writes are happening -- and not the gaps in between, where reads are
+permitted.
+
+As you might have surmised, these sorts of "discontinuous" borrows
+represent a kind of "step up" in the complexity of the system. If it
+were vital to accept examples with interleaved writes like the
+previous one, then this wouldn't bother me (NLL also represents such a
+step, for example, but it seems clearly worth it). But given that the
+example is artificial and not a pattern I have ever seen arise in
+"real life", it seems like we should try to avoid growing the
+underlying complexity of the system if we can.
+
+To see what I mean about a "step up" in complexity, consider how we
+would integrate this proposal into lifetime inference. The current
+rules treat all regions equally, but this proposal seems to imply that
+regions have "roles".  For example, the `'r` region captures the
+"liveness" constraints that I described in the original NLL
+proposal. Meanwhile the `'w` region captures "activity".
+
+(Since we would always convert a `&'a mut T` type into `RefMut<'a, 'a,
+T>`, all regions in struct parameters would adopt the more
+conservative "liveness" role to start. This is good because we
+wouldn't want to start allowing "holes" in the lifetimes that unsafe
+code is relying on to prevent access from the outside. It would
+however be possible for type inference to use a `RefMut<'r, 'w ,T>`
+type as the value for a type parameter; I don't yet see a way for that
+to cause any surprises, but perhaps it can if you consider
+specialization and other non-parametric features.)
+
+Another example of where this "complexity step" surfaces came from
+[Ralf Jung][rjung]. As you may know, Ralf is working on a
+formalization of Rust as part of the [RustBelt project][rb] (if you're
+interested, there is video available of a
+[great introduction to this work][am] which Ralf gave at the Rust
+Paris meetup). In any case, their model is a kind of generalization of
+Rust, in that it can accept a lot of programs that standard Rust
+cannot (it is intended to be used for assigning types to unsafe code
+as well as safe code). The two-phase borrow proposal that I describe
+here should be able to fit into that system in a fairly
+straightforward way. But if we adopted discontinuous regions, that
+would require making Ralf's system more expressive. This is not
+necessarily an argument against doing it, but it does show that it
+makes the Rust system qualitatively more complex to reason about.
```

Comment:
> The two-phase borrow proposal that I describe
+here should be able to fit into that system in a fairly
+straightforward way.

I like your optimism here. ;)